File: src/content/3.11/code/scala/snippet08.scala
def fst[I]: ((I, _)) => I = _._1

def toExp[A, B]: (A => B) => Exp[A, B] = f => new Lan[(A, ?), I, B] {
  def fk[L](ki: (A, L)): B =
    f.compose(fst[A])(ki)

  def di[L]: I[L] = I()
}

def fromExp[A, B]: Exp[A, B] => (A => B) =
  lan => a => lan.fk((a, lan.di))

File: src/content/3.11/code/ocaml/snippet08.ml
let to_exp (type a' b') f =
  (module struct
    type a = a'
    type b = b'
    type 'a d = I of 'a
    type 'a k = 'a * a
    type i = unit

    let fk (a, _) = f a
    let di = I ()
  end : Exp
    with type a = a'
     and type b = b')
;;

let from_exp
    (type a' b')
    (module E : Exp with type a = a' and type b = b')
    a
  =
  let (I i) = E.di in
  E.fk (a, i)
;;
